src/coreclr/gc/gc.cpp

```diff
@@ -3710,44 +3721,75 @@ bool region_allocator::allocate_region (size_t size, uint8_t** start, uint8_t**
 }
 
-void region_allocator::delete_region (uint8_t* start)
+void region_allocator::delete_region (uint8_t* region_start)
 {
     enter_spin_lock();
 
-    assert (is_region_aligned (start));
+    assert (is_region_aligned (region_start));
 
     print_map ("before delete");
 
-    uint32_t* current_index = region_map_index_of (start);
+    uint32_t* current_index = region_map_index_of (region_start);
     uint32_t current_val = *current_index;
     assert (!is_unit_memory_free (current_val));
 
-    dprintf (REGIONS_LOG, ("----DEL %d-----", (current_index - region_map_start)));
+    dprintf (REGIONS_LOG, ("----DEL %d-----", (current_index - region_map_left_start)));
+    uint32_t* region_end_index = current_index + current_val;
+    uint8_t* region_end = region_address_of (region_end_index);
 
-    uint8_t* region_end = region_address_of (current_index + current_val);
-    if (region_end == global_region_used)
+    int free_block_size = current_val;
+    uint32_t* free_index = current_index;
+    if ((current_index != region_map_left_start) && (current_index != region_map_right_start))
     {
-        region_map_end = current_index;
-        dprintf (REGIONS_LOG, ("adjust global used from %Ix to %Ix", 
-            global_region_used, region_address_of (current_index)));
-        global_region_used = region_address_of (current_index);
+        uint32_t previous_val = *(current_index - 1);
+        if (is_unit_memory_free(previous_val))
+        {
+            uint32_t previous_size = get_num_units (previous_val);
+            free_index -= previous_size;
+            free_block_size += previous_size;
+        }
+    }
+    if ((region_end != global_region_left_used) && (region_end != global_region_end))
+    {
+        uint32_t next_val = *region_end_index;
+        if (is_unit_memory_free(next_val))
+        {
+            uint32_t next_size = get_num_units (next_val);
+            free_block_size += next_size;
+            region_end += next_size;
+        }
+    }
+    if (region_end == global_region_left_used)
+    {
+        region_map_left_end = free_index;
+        dprintf (REGIONS_LOG, ("adjust global left used from %Ix to %Ix", 
+            global_region_left_used, region_address_of (free_index)));
+        global_region_left_used = region_address_of (free_index);
+    }
+    else if (region_start == global_region_right_used)
+    {
+        region_map_right_start = free_index + free_block_size;
+        dprintf (REGIONS_LOG, ("adjust global right used from %Ix to %Ix", 
+            global_region_right_used, region_address_of (free_index + free_block_size)));
+        global_region_right_used = region_address_of (free_index + free_block_size);
     }
     else
     {
-        make_free_block (current_index, current_val);
+        make_free_block (free_index, free_block_size);
     }
 
-    total_free_units += current_val;
+    total_free_units += free_block_size;
```

Comment:
you are double counting here if the free_block_size was already increased to include its previous and/or next free block size.